

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>lncrnapy.algorithms &mdash; lncRNA-Py  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../_static/doctools.js?v=9a2dae69"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="lncrnapy.selection" href="selection.html" />
    <link rel="prev" title="lncrnapy.utils" href="utils.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            lncRNA-Py
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../scripts.html">Running Scripts</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../api.html">API Documentation</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="data.html">lncrnapy.data</a></li>
<li class="toctree-l2"><a class="reference internal" href="features.html">lncrnapy.features</a></li>
<li class="toctree-l2"><a class="reference internal" href="modules.html">lncrnapy.modules</a></li>
<li class="toctree-l2"><a class="reference internal" href="train.html">lncrnapy.train</a></li>
<li class="toctree-l2"><a class="reference internal" href="evaluate.html">lncrnapy.evaluate</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html">lncrnapy.utils</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">lncrnapy.algorithms</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#module-lncrnapy.algorithms.algorithm">Algorithm</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#lncrnapy.algorithms.algorithm.Algorithm"><code class="docutils literal notranslate"><span class="pre">Algorithm</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-lncrnapy.algorithms.traditional">Traditional</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#lncrnapy.algorithms.traditional.CNCI"><code class="docutils literal notranslate"><span class="pre">CNCI</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#lncrnapy.algorithms.traditional.CNIT"><code class="docutils literal notranslate"><span class="pre">CNIT</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#lncrnapy.algorithms.traditional.CPAT"><code class="docutils literal notranslate"><span class="pre">CPAT</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#lncrnapy.algorithms.traditional.CPC"><code class="docutils literal notranslate"><span class="pre">CPC</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#lncrnapy.algorithms.traditional.CPC2"><code class="docutils literal notranslate"><span class="pre">CPC2</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#lncrnapy.algorithms.traditional.CPPred"><code class="docutils literal notranslate"><span class="pre">CPPred</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#lncrnapy.algorithms.traditional.DeepCPP"><code class="docutils literal notranslate"><span class="pre">DeepCPP</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#lncrnapy.algorithms.traditional.FEELnc"><code class="docutils literal notranslate"><span class="pre">FEELnc</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#lncrnapy.algorithms.traditional.LncADeep"><code class="docutils literal notranslate"><span class="pre">LncADeep</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#lncrnapy.algorithms.traditional.LncFinder"><code class="docutils literal notranslate"><span class="pre">LncFinder</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#lncrnapy.algorithms.traditional.PLEK"><code class="docutils literal notranslate"><span class="pre">PLEK</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#lncrnapy.algorithms.traditional.PLncPro"><code class="docutils literal notranslate"><span class="pre">PLncPro</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#lncrnapy.algorithms.traditional.iSeeRNA"><code class="docutils literal notranslate"><span class="pre">iSeeRNA</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="selection.html">lncrnapy.selection</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">lncRNA-Py</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../api.html">API Documentation</a></li>
      <li class="breadcrumb-item active">lncrnapy.algorithms</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/luukromeijn/rhythmnblues/blob/master/docs/api_docs/algorithms.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-lncrnapy.algorithms">
<span id="lncrnapy-algorithms"></span><h1>lncrnapy.algorithms<a class="headerlink" href="#module-lncrnapy.algorithms" title="Link to this heading"></a></h1>
<section id="module-lncrnapy.algorithms.algorithm">
<span id="algorithm"></span><h2>Algorithm<a class="headerlink" href="#module-lncrnapy.algorithms.algorithm" title="Link to this heading"></a></h2>
<p>Contains <cite>Algorithm</cite> base class for the classificiation of RNA transcritps as
either protein-coding or long non-coding.</p>
<dl class="py class">
<dt class="sig sig-object py" id="lncrnapy.algorithms.algorithm.Algorithm">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">lncrnapy.algorithms.algorithm.</span></span><span class="sig-name descname"><span class="pre">Algorithm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">feature_extractors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">used_features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#lncrnapy.algorithms.algorithm.Algorithm" title="Link to this definition"></a></dt>
<dd><p>Base class for algorithms for the classification of RNA transcripts as
either protein-coding or long non-coding.</p>
<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">`model`</span></span></dt>
<dd><p>Underlying classification model. Can be a trained <cite>torch.nn.Module</cite>
object with a single, sigmoid-activated output node (lncrnapy
<cite>Model</cite> recommended.), or a scikit-learn-style model with a <cite>.fit</cite> and
<cite>.classify</cite> method.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">`feature_extractors`</span></span></dt>
<dd><p>Feature extractor or list of feature extractors that are applied to the
data if a feature in <cite>used_features</cite> is missing in the input.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><cite>list</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">`used_features`</span></span></dt>
<dd><p>Specifies which feature names (data columns) serve as input variables
for the model. If None, will use all features from <cite>feature_extractors</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><cite>list[str]</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="lncrnapy.algorithms.algorithm.Algorithm.feature_extraction">
<span class="sig-name descname"><span class="pre">feature_extraction</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#lncrnapy.algorithms.algorithm.Algorithm.feature_extraction" title="Link to this definition"></a></dt>
<dd><p>Calls upon the object’s feature extractors if a feature in the
<cite>used_features</cite> attribute is missing in <cite>data</cite>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="lncrnapy.algorithms.algorithm.Algorithm.fit">
<span class="sig-name descname"><span class="pre">fit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#lncrnapy.algorithms.algorithm.Algorithm.fit" title="Link to this definition"></a></dt>
<dd><p>Fits model on <cite>data</cite>, extracting features first if necessary. Will
only fit on features as specified in the <cite>used_features</cite> attribute.
This method is disabled when the <cite>model</cite> attribute is a
<cite>torch.nn.Module</cite> instance.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="lncrnapy.algorithms.algorithm.Algorithm.predict">
<span class="sig-name descname"><span class="pre">predict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#lncrnapy.algorithms.algorithm.Algorithm.predict" title="Link to this definition"></a></dt>
<dd><p>Classifies <cite>data</cite>, extracting features first if necessary. Will
only use features as specified in the <cite>used_features</cite> attribute.</p>
</dd></dl>

</dd></dl>

</section>
<section id="module-lncrnapy.algorithms.traditional">
<span id="traditional"></span><h2>Traditional<a class="headerlink" href="#module-lncrnapy.algorithms.traditional" title="Link to this heading"></a></h2>
<p>Pre-implemented coding/non-coding RNA classifiers.</p>
<p>Note that all classifiers that are based on algorithms presented in related
works should be considered as loose adaptations. We do not guarantee that our
implementations achieve the exact same performance as that of the original
works.</p>
<dl class="py class">
<dt class="sig sig-object py" id="lncrnapy.algorithms.traditional.CNCI">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">lncrnapy.algorithms.traditional.</span></span><span class="sig-name descname"><span class="pre">CNCI</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ant_ref</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#lncrnapy.algorithms.traditional.CNCI" title="Link to this definition"></a></dt>
<dd><p>Coding Non-Coding Index (CNCI)</p>
<p class="rubric">References</p>
<p>CNCI: Sun et al. (2013) <a class="reference external" href="https://doi.org/10.1093/nar/gkt646">https://doi.org/10.1093/nar/gkt646</a></p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="lncrnapy.algorithms.traditional.CNIT">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">lncrnapy.algorithms.traditional.</span></span><span class="sig-name descname"><span class="pre">CNIT</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ant_ref</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#lncrnapy.algorithms.traditional.CNIT" title="Link to this definition"></a></dt>
<dd><p>Coding-Non-Coding Identifying Tool (CNIT)</p>
<p class="rubric">References</p>
<p>CNIT: Guo et al. (2019) <a class="reference external" href="https://doi.org/10.1093/nar/gkz400">https://doi.org/10.1093/nar/gkz400</a></p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="lncrnapy.algorithms.traditional.CPAT">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">lncrnapy.algorithms.traditional.</span></span><span class="sig-name descname"><span class="pre">CPAT</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fickett_ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hexamer_ref</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#lncrnapy.algorithms.traditional.CPAT" title="Link to this definition"></a></dt>
<dd><p>Coding-Potential Assesment Tool (CPAT).</p>
<p class="rubric">References</p>
<p>CPAT: Wang et al. (2013) <a class="reference external" href="https://doi.org/10.1093/nar/gkt006">https://doi.org/10.1093/nar/gkt006</a></p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="lncrnapy.algorithms.traditional.CPC">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">lncrnapy.algorithms.traditional.</span></span><span class="sig-name descname"><span class="pre">CPC</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">database</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#lncrnapy.algorithms.traditional.CPC" title="Link to this definition"></a></dt>
<dd><p>Adaptation of Coding Potential Calculator (CPC). This adaptation differs
in two ways from the original: 1) we use ORF length instead of the log-odds
score; 2) we do not make us of the ORF integrity feature, as all identified
ORFs will have a start and stop codon.</p>
<p class="rubric">References</p>
<p>CPC: Kong et al. (2007) <a class="reference external" href="https://doi.org/10.1093/nar/gkm391">https://doi.org/10.1093/nar/gkm391</a></p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="lncrnapy.algorithms.traditional.CPC2">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">lncrnapy.algorithms.traditional.</span></span><span class="sig-name descname"><span class="pre">CPC2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fickett_ref</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#lncrnapy.algorithms.traditional.CPC2" title="Link to this definition"></a></dt>
<dd><p>Adaptation of Coding Potential Calculator version 2 (CPC2). An important
difference between this implementation and the original is that we do not
make us of the ORF integrity feature, as all ORFs will have a start and stop
codon.</p>
<p class="rubric">References</p>
<p>CPC2: Kang et al. (2017) <a class="reference external" href="https://doi.org/10.1093/nar/gkx428">https://doi.org/10.1093/nar/gkx428</a></p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="lncrnapy.algorithms.traditional.CPPred">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">lncrnapy.algorithms.traditional.</span></span><span class="sig-name descname"><span class="pre">CPPred</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fickett_ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hexamer_ref</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#lncrnapy.algorithms.traditional.CPPred" title="Link to this definition"></a></dt>
<dd><p>Adaptation of Coding Potential Prediction (CPPred). We replace C and T
of CTD features with mono- and dimer frequencies, respectively.</p>
<p class="rubric">References</p>
<p>CPPred: Tong et al. (2019) <a class="reference external" href="https://doi.org/10.1093/nar/gkz087">https://doi.org/10.1093/nar/gkz087</a></p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="lncrnapy.algorithms.traditional.DeepCPP">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">lncrnapy.algorithms.traditional.</span></span><span class="sig-name descname"><span class="pre">DeepCPP</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fickett_ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hexamer_ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zhang_ref</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#lncrnapy.algorithms.traditional.DeepCPP" title="Link to this definition"></a></dt>
<dd><p>Deep neural network for coding potential prediction (DeepCPP).</p>
<p class="rubric">References</p>
<p>DeepCPP: Zhang et al. (2020) <a class="reference external" href="https://doi.org/10.1093/bib/bbaa039">https://doi.org/10.1093/bib/bbaa039</a></p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="lncrnapy.algorithms.traditional.FEELnc">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">lncrnapy.algorithms.traditional.</span></span><span class="sig-name descname"><span class="pre">FEELnc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kmer_refs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#lncrnapy.algorithms.traditional.FEELnc" title="Link to this definition"></a></dt>
<dd><p>FlExible Extraction of LncRNAs (FEELnc)</p>
<p class="rubric">References</p>
<p>FEELnc: Wucher et al. (2017) <a class="reference external" href="https://doi.org/10.1093/nar/gkw1306">https://doi.org/10.1093/nar/gkw1306</a></p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="lncrnapy.algorithms.traditional.LncADeep">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">lncrnapy.algorithms.traditional.</span></span><span class="sig-name descname"><span class="pre">LncADeep</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fickett_ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hexamer_ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">database</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#lncrnapy.algorithms.traditional.LncADeep" title="Link to this definition"></a></dt>
<dd><p>Adaptation of the LncADeep algorithm, a feature-based classifier based on
a Deep Belief Network. We replace HMMER with BLASTX, and implement the
full-length variant as explained in the paper.</p>
<p class="rubric">References</p>
<p>LncADeep: Yang et al. (2018) <a class="reference external" href="https://doi.org/10.1093/bioinformatics/bty428">https://doi.org/10.1093/bioinformatics/bty428</a></p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="lncrnapy.algorithms.traditional.LncFinder">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">lncrnapy.algorithms.traditional.</span></span><span class="sig-name descname"><span class="pre">LncFinder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">orf_6mer_ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">acguD_4mer_ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">acguACGU_3mer_ref</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#lncrnapy.algorithms.traditional.LncFinder" title="Link to this definition"></a></dt>
<dd><p>LncFinder algorithm. Utilizes log distance of (ORF) k-mer frequencies, as
well as secondary structure elements and EIIP-derived physico-chemical
features.</p>
<p class="rubric">References</p>
<p>LncFinder: Han et al. (2018) <a class="reference external" href="https://doi.org/10.1093/bib/bby065">https://doi.org/10.1093/bib/bby065</a></p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="lncrnapy.algorithms.traditional.PLEK">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">lncrnapy.algorithms.traditional.</span></span><span class="sig-name descname"><span class="pre">PLEK</span></span><a class="headerlink" href="#lncrnapy.algorithms.traditional.PLEK" title="Link to this definition"></a></dt>
<dd><p>Predictor of long non-coding RNAs and messenger RNAs based on an improved
k-mer scheme (PLEK)</p>
<p class="rubric">References</p>
<p>PLEK: Li et al. (2014) <a class="reference external" href="https://doi.org/10.1186/1471-2105-15-311">https://doi.org/10.1186/1471-2105-15-311</a></p>
<dl class="py method">
<dt class="sig sig-object py" id="lncrnapy.algorithms.traditional.PLEK.feature_extraction">
<span class="sig-name descname"><span class="pre">feature_extraction</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#lncrnapy.algorithms.traditional.PLEK.feature_extraction" title="Link to this definition"></a></dt>
<dd><p>Calls upon the object’s feature extractors if a feature in the
<cite>used_features</cite> attribute is missing in <cite>data</cite>.</p>
<p>Altered for PLEK algorithm to prevent requiring special PLEK nucleotide
frequencies (will calculate them instead).</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="lncrnapy.algorithms.traditional.PLncPro">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">lncrnapy.algorithms.traditional.</span></span><span class="sig-name descname"><span class="pre">PLncPro</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">database</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#lncrnapy.algorithms.traditional.PLncPro" title="Link to this definition"></a></dt>
<dd><p>Plant Long Non-Coding RNA Prediction by Random fOrest</p>
<p class="rubric">References</p>
<p>PLncPro: Singh et al. (2017) <a class="reference external" href="https://doi.org/10.1093/nar/gkx866">https://doi.org/10.1093/nar/gkx866</a></p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="lncrnapy.algorithms.traditional.iSeeRNA">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">lncrnapy.algorithms.traditional.</span></span><span class="sig-name descname"><span class="pre">iSeeRNA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">database</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#lncrnapy.algorithms.traditional.iSeeRNA" title="Link to this definition"></a></dt>
<dd><p>Adaptation of the iSeeRNA algorithm. An important difference is that the
conservation score feature is replaced by the number of BLASTX hits.</p>
<p class="rubric">References</p>
<p>iSeeRNA: Sun et al. (2013) <a class="reference external" href="https://doi.org/10.1186/1471-2164-14-S2-S7">https://doi.org/10.1186/1471-2164-14-S2-S7</a></p>
</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="utils.html" class="btn btn-neutral float-left" title="lncrnapy.utils" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="selection.html" class="btn btn-neutral float-right" title="lncrnapy.selection" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Luuk Romeijn.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>