

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>LncRNA classification with an RNA language model &mdash; lncRNA-Py  documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="_static/nbsphinx-code-cells.css?v=2aa19091" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=5929fcd5"></script>
      <script src="_static/doctools.js?v=9a2dae69"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
      <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
      <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="lncrnapy.selection" href="api_docs/selection.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            lncRNA-Py
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="scripts.html">Running Scripts</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="api.html">API Documentation</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="api_docs/data.html">lncrnapy.data</a></li>
<li class="toctree-l2"><a class="reference internal" href="api_docs/features.html">lncrnapy.features</a></li>
<li class="toctree-l2"><a class="reference internal" href="api_docs/modules.html">lncrnapy.modules</a></li>
<li class="toctree-l2"><a class="reference internal" href="api_docs/train.html">lncrnapy.train</a></li>
<li class="toctree-l2"><a class="reference internal" href="api_docs/evaluate.html">lncrnapy.evaluate</a></li>
<li class="toctree-l2"><a class="reference internal" href="api_docs/utils.html">lncrnapy.utils</a></li>
<li class="toctree-l2"><a class="reference internal" href="api_docs/algorithms.html">lncrnapy.algorithms</a></li>
<li class="toctree-l2"><a class="reference internal" href="api_docs/selection.html">lncrnapy.selection</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">LncRNA classification with an RNA language model</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Data">Data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Encoding">Encoding</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Neural-architecture">Neural architecture</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Pre-training">Pre-training</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Fine-tuning">Fine-tuning</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">lncRNA-Py</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="api.html">API Documentation</a></li>
      <li class="breadcrumb-item active">LncRNA classification with an RNA language model</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/luukromeijn/rhythmnblues/blob/master/docs/example.ipynb" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="LncRNA-classification-with-an-RNA-language-model">
<h1>LncRNA classification with an RNA language model<a class="headerlink" href="#LncRNA-classification-with-an-RNA-language-model" title="Link to this heading"></a></h1>
<p>This example notebook shows the basic functionalities of the <code class="docutils literal notranslate"><span class="pre">lncrnapy</span></code> package and how it can be used to train an RNA language model for lncRNA classification.</p>
<section id="Data">
<h2>Data<a class="headerlink" href="#Data" title="Link to this heading"></a></h2>
<p>Let us start by loading some sequence data. The <code class="docutils literal notranslate"><span class="pre">Data</span></code> object accepts either a single fasta file (for unlabelled data), or a list of two fasta files (for pcRNA and lncRNA, respectively).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from lncrnapy.data import Data

data_dir = &#39;/data/s2592800/data&#39; # Change this for your setup

pretrain_data = Data([f&#39;{data_dir}/sequences/pretrain_human_pcrna.fasta&#39;,
                      f&#39;{data_dir}/sequences/pretrain_human_ncrna.fasta&#39;])

print(pretrain_data)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Importing data...
Imported 297724 protein-coding and 238470 non-coding RNA transcripts with 0 feature(s).
                                                       id  \
0       ENST00000676272.1|ENSG00000087053.20|OTTHUMG00...
1       ENST00000676132.1|ENSG00000133424.22|OTTHUMG00...
2       ENST00000504953.5|ENSG00000196104.11|OTTHUMG00...
3       ENST00000677479.1|ENSG00000168610.17|OTTHUMG00...
4       ENST00000513066.3|ENSG00000113407.14|OTTHUMG00...
...                                                   ...
536189                                        NR_026711.1
536190                                        NR_189643.1
536191                                        NR_026710.1
536192                                        NR_027231.1
536193                                        NR_130733.1

                                                 sequence  label
0       AGCCTACAGGCGCGGTGCACTCTGGGGGAACATGGCCGCTTCCGGT...  pcRNA
1       AAGGATCCTCATGGCAGCATGGAGAGGCCAGGCGTGTCTGCGTGGC...  pcRNA
2       TCCTCTGGCTCCAGCCAAGCGTCCTATCCGGAGCCAACTGTAGCTG...  pcRNA
3       GCAGACTGGGAGGGGGAGCCGGGGGTTCCGACGTCGCAGCCGAGGG...  pcRNA
4       TCCCGGGCGCTAGCCCACCTCCCACCCGCCTCTTGGCTCCTCTCCT...  pcRNA
...                                                   ...    ...
536189  TTGTTTTCTTTTATTTATTTATTTAGACGGAGTCTCACTCTGTCGC...  ncRNA
536190  CTAGGAGCCAAGACCTCTTGCTGGCTGCCACATCGGCACTGCCAAC...  ncRNA
536191  TTGTTTTCTTTTATTTATTTATTTAGACGGAGTCTCACTCTGTCGC...  ncRNA
536192  GGGACCGGGCCGCTTCTTGTTCGGCGTGGGCGGCTCCTGGCAGACC...  ncRNA
536193  AAGCATTTGGCAAGCACATCCTGGGTGCCAGGCCCAAGCAGGCTGG...  ncRNA

[536194 rows x 3 columns]
</pre></div></div>
</div>
</section>
<section id="Encoding">
<h2>Encoding<a class="headerlink" href="#Encoding" title="Link to this heading"></a></h2>
<p>Next, we encode the data into a numeric Tensor format that is compatible with the neural network that we wish to use. For Convolutional Sequence Encoding, we must encode the data into a four-dimensional representation.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pretrain_data.set_tensor_features(&#39;4D-DNA&#39;)
</pre></div>
</div>
</div>
<p>To illustrate what we just did, let us sample a sequence by indexing.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sequence, label = pretrain_data[0] # Sample the first sequence
print(&quot;Sequence:&quot;)
print(sequence)
print(&quot;Label (0=ncRNA, 1=pcRNA):&quot;)
print(label)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Sequence:
tensor([[1., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 1.,  ..., 0., 0., 0.],
        [0., 1., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.]], device=&#39;cuda:0&#39;)
Label (0=ncRNA, 1=pcRNA):
tensor([1.], device=&#39;cuda:0&#39;)
</pre></div></div>
</div>
</section>
<section id="Neural-architecture">
<h2>Neural architecture<a class="headerlink" href="#Neural-architecture" title="Link to this heading"></a></h2>
<p>Now we can define a model. The most important component of our model is its base architecture. <code class="docutils literal notranslate"><span class="pre">lncrnapy.modules</span></code> contains several types of architectures implementations, including CNNs (e.g. <code class="docutils literal notranslate"><span class="pre">ResNet</span></code>) and language models (e.g. <code class="docutils literal notranslate"><span class="pre">BERT</span></code>). Convolutional Sequence Encoding requires a special variant of BERT, which is implemented in the <code class="docutils literal notranslate"><span class="pre">CSEBERT</span></code> class.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from lncrnapy.modules import CSEBERT
base_arch = CSEBERT(n_kernels=768, kernel_size=9)
</pre></div>
</div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">lncrnapy.modules</span></code> also contains several wrapper classes that encapsulate a base architecture and add the required layers to perform tasks like classification, regression, and masked language modeling. For example, this is how we turn our model into a classifier:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from lncrnapy.modules import Classifier
from lncrnapy import utils

model = Classifier(base_arch)
model = model.to(utils.DEVICE) # Send the model to the GPU.
</pre></div>
</div>
</div>
<p>Now let’s make a prediction on the validation dataset.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># Load and encode the data
valid_data = Data([f&#39;{data_dir}/sequences/valid_gencode_pcrna.fasta&#39;,
                   f&#39;{data_dir}/sequences/valid_gencode_ncrna.fasta&#39;])
valid_data.set_tensor_features(&#39;4D-DNA&#39;)

# Make a prediction
prediction = model.predict(valid_data)
print(prediction)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Importing data...
Imported 5583 protein-coding and 2998 non-coding RNA transcripts with 0 feature(s).
tensor([[0.4755],
        [0.4519],
        [0.4753],
        ...,
        [0.4651],
        [0.4455],
        [0.4553]])
</pre></div></div>
</div>
</section>
<section id="Pre-training">
<h2>Pre-training<a class="headerlink" href="#Pre-training" title="Link to this heading"></a></h2>
<p>The prediction made above is only a random one, as we have not trained our model yet. Language models are often pre-trained before being fine-tuned to perform a specific task. We shall do the same.</p>
<p>First, we must wrap the base architecture into the proper wrapper class:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from lncrnapy.modules import MaskedConvModel

model = MaskedConvModel(base_arch).to(utils.DEVICE)
</pre></div>
</div>
</div>
<p>Now we can train it using the <code class="docutils literal notranslate"><span class="pre">train_masked_conv_modeling</span></code> function.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from lncrnapy.train import train_masked_conv_modeling

model, history = train_masked_conv_modeling(
    model, pretrain_data, valid_data, epochs=500
)

print(history) # Contains the performance at every epoch
</pre></div>
</div>
</div>
</section>
<section id="Fine-tuning">
<h2>Fine-tuning<a class="headerlink" href="#Fine-tuning" title="Link to this heading"></a></h2>
<p>After pre-training, we can extract the base architecture and wrap it inside a <code class="docutils literal notranslate"><span class="pre">Classifier</span></code> object again. We can then fine-tune our model using the <code class="docutils literal notranslate"><span class="pre">train_classifier</span></code> function.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from lncrnapy.train import train_classifier

finetune_data = Data([f&#39;{data_dir}/sequences/finetune_gencode_pcrna.fasta&#39;,
                      f&#39;{data_dir}/sequences/finetune_gencode_ncrna.fasta&#39;])

model = Classifier(model.base_arch).to(utils.DEVICE)
model, history = train_classifier(model, pretrain_data, valid_data, epochs=100)

print(history)
</pre></div>
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="api_docs/selection.html" class="btn btn-neutral float-left" title="lncrnapy.selection" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Luuk Romeijn.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>