

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>&lt;no title&gt; &mdash; lncRNA-Py  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../_static/doctools.js?v=9a2dae69"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            lncRNA-Py
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../scripts.html">Running Scripts</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html">API Documentation</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">lncRNA-Py</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">&lt;no title&gt;</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/luukromeijn/rhythmnblues/blob/master/docs/script_args/pretrain_args.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <dl>
<dt>usage: get_script_args.py [-h] [–exp_prefix EXP_PREFIX] [–encoding_method {conv,bpe,kmer,nuc}] [–epochs EPOCHS] [–n_samples_per_epoch N_SAMPLES_PER_EPOCH] [–batch_size BATCH_SIZE] [–warmup_steps WARMUP_STEPS] [–d_model D_MODEL] [–N N] [–d_ff D_FF] [–h H] [–dropout DROPOUT] [–n_kernels N_KERNELS] [–kernel_size KERNEL_SIZE] [–bpe_file BPE_FILE] [–k K] [–p_mlm P_MLM] [–p_mask P_MASK] [–p_random P_RANDOM]</dt><dd><p>[–context_length CONTEXT_LENGTH] [–data_dir DATA_DIR] [–results_dir RESULTS_DIR] [–model_dir MODEL_DIR] [–mask_size MASK_SIZE] [–no_random_reading_frame] [–input_linear] [–no_input_relu] [–no_output_linear] [–output_relu]
fasta_train fasta_valid</p>
</dd>
<dt>positional arguments:</dt><dd><p>fasta_train           Path to FASTA file with pre-training sequences. (str)
fasta_valid           Path to FASTA file with sequences to use for validating model performance after every epoch. (str)</p>
</dd>
<dt>optional arguments:</dt><dd><dl class="option-list">
<dt><kbd><span class="option">-h</span>, <span class="option">--help</span></kbd></dt>
<dd><p>show this help message and exit</p>
</dd>
<dt><kbd><span class="option">--exp_prefix <var>EXP_PREFIX</var></span></kbd></dt>
<dd><p>Added prefix to model/experiment name. (str=”MLM”)</p>
</dd>
</dl>
<dl class="simple">
<dt>–encoding_method {conv,bpe,kmer,nuc}</dt><dd><p>Sequence encoding method. (str=”conv”)</p>
</dd>
</dl>
<dl class="option-list">
<dt><kbd><span class="option">--epochs <var>EPOCHS</var></span></kbd></dt>
<dd><p>Number of epochs to pre-train for. (int=500)</p>
</dd>
<dt><kbd><span class="option">--n_samples_per_epoch <var>N_SAMPLES_PER_EPOCH</var></span></kbd></dt>
<dd><p>Number of training samples per epoch. (int=10000)</p>
</dd>
<dt><kbd><span class="option">--batch_size <var>BATCH_SIZE</var></span></kbd></dt>
<dd><p>Number of samples per optimization step. (int=8)</p>
</dd>
<dt><kbd><span class="option">--warmup_steps <var>WARMUP_STEPS</var></span></kbd></dt>
<dd><p>Number of optimization steps in which learning rate increases linearly. After this amount of steps, the learning rate decreases proportional to the inverse square root of the step number. (int=8)</p>
</dd>
<dt><kbd><span class="option">--d_model <var>D_MODEL</var></span></kbd></dt>
<dd><p>BERT embedding dimensionality. (int=768)</p>
</dd>
<dt><kbd><span class="option">--N <var>N</var></span></kbd></dt>
<dd><p>Number of BERT transformer blocks. (int=12)</p>
</dd>
<dt><kbd><span class="option">--d_ff <var>D_FF</var></span></kbd></dt>
<dd><p>Number of nodes in BERT FFN sublayers (int=4*d_model)</p>
</dd>
<dt><kbd><span class="option">--h <var>H</var></span></kbd></dt>
<dd><p>Number of BERT self-attention heads (int=int(d_model/64))</p>
</dd>
<dt><kbd><span class="option">--dropout <var>DROPOUT</var></span></kbd></dt>
<dd><p>Dropout probability in MLM output head. (float=0)</p>
</dd>
<dt><kbd><span class="option">--n_kernels <var>N_KERNELS</var></span></kbd></dt>
<dd><p>Specifies number of kernels when convolutional sequence encoding is used. (int=768)</p>
</dd>
<dt><kbd><span class="option">--kernel_size <var>KERNEL_SIZE</var></span></kbd></dt>
<dd><p>Specifies kernel size when convolutional sequence encoding is used. (int=10)</p>
</dd>
<dt><kbd><span class="option">--bpe_file <var>BPE_FILE</var></span></kbd></dt>
<dd><p>Filepath to BPE model generated with BPE script. Required when Byte Pair Encoding is used. (str=””)</p>
</dd>
<dt><kbd><span class="option">--k <var>K</var></span></kbd></dt>
<dd><p>Specifies k when k-mer encoding is used. (int=6)</p>
</dd>
<dt><kbd><span class="option">--p_mlm <var>P_MLM</var></span></kbd></dt>
<dd><p>Selection probability per token/nucleotide in MLM. (float=0.15)</p>
</dd>
<dt><kbd><span class="option">--p_mask <var>P_MASK</var></span></kbd></dt>
<dd><p>Mask probability for selected token/nucleotide. (float=0.8)</p>
</dd>
<dt><kbd><span class="option">--p_random <var>P_RANDOM</var></span></kbd></dt>
<dd><p>Random replacement chance per token/nucleotide. (float=0.1)</p>
</dd>
<dt><kbd><span class="option">--context_length <var>CONTEXT_LENGTH</var></span></kbd></dt>
<dd><p>Number of input positions. For cse/k-mer encoding, this translates to a maximum of (768-1)*k input nucleotides. (int=768)</p>
</dd>
<dt><kbd><span class="option">--data_dir <var>DATA_DIR</var></span></kbd></dt>
<dd><p>Parent directory to use for any of the paths specified in these arguments. (str=””)</p>
</dd>
<dt><kbd><span class="option">--results_dir <var>RESULTS_DIR</var></span></kbd></dt>
<dd><p>Parent directory to use for the results folder of this script. (str=””)</p>
</dd>
<dt><kbd><span class="option">--model_dir <var>MODEL_DIR</var></span></kbd></dt>
<dd><p>Directory where to save pre-trained model to. Model with the highest accuracy on the validation dataset is saved. (str=f”{data_dir}/models”)</p>
</dd>
<dt><kbd><span class="option">--mask_size <var>MASK_SIZE</var></span></kbd></dt>
<dd><p>Number of contiguous nucleotides that make up a mask. (int=1)</p>
</dd>
<dt><kbd><span class="option">--no_random_reading_frame</span></kbd></dt>
<dd><p>Turns off sampling in random reading frame for convolutional sequence encoding (bool)</p>
</dd>
<dt><kbd><span class="option">--input_linear</span></kbd></dt>
<dd><p>Forces linear projection of kernels onto d_model dimensions in convolutional sequence encoding. (bool)</p>
</dd>
<dt><kbd><span class="option">--no_input_relu</span></kbd></dt>
<dd><p>Turns off ReLU activation of kernels in convolutional sequence encoding. (bool)</p>
</dd>
<dt><kbd><span class="option">--no_output_linear</span></kbd></dt>
<dd><p>Forces linear projection of embeddings onto n_kernels dimensions before masked convolution output layer. (bool)</p>
</dd>
<dt><kbd><span class="option">--output_relu</span></kbd></dt>
<dd><p>Forces ReLU activation of embeddings before masked convolution output layer. (bool)</p>
</dd>
</dl>
</dd>
</dl>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Luuk Romeijn.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>