

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Running Scripts &mdash; lncRNA-Py  documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=5929fcd5"></script>
      <script src="_static/doctools.js?v=9a2dae69"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="API Documentation" href="api.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            lncRNA-Py
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Running Scripts</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#classify">Classify</a></li>
<li class="toctree-l2"><a class="reference internal" href="#train">Train</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pre-train">Pre-Train</a></li>
<li class="toctree-l2"><a class="reference internal" href="#generate-embeddings">Generate Embeddings</a></li>
<li class="toctree-l2"><a class="reference internal" href="#byte-pair-encoding">Byte Pair Encoding</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="experiments.html">Experiments</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">lncRNA-Py</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Running Scripts</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/luukromeijn/rhythmnblues/blob/master/docs/scripts.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="running-scripts">
<h1>Running Scripts<a class="headerlink" href="#running-scripts" title="Link to this heading"></a></h1>
<p>Scripts for using lncRNA-BERT for classification and generating embeddings, as
well as for (pre-)training and fine-tuning custom models. For feature-based
classification or more advanced use-cases, users can write their own custom
scripts using the <a class="reference external" href="api.html">API</a>. An example notebook is provided
<a class="reference external" href="https://colab.research.google.com/drive/1NSsFYvQQbwhH0yf7wEVfjxyvqG-bUrUS?usp=sharing">here</a>.</p>
<section id="classify">
<h2>Classify<a class="headerlink" href="#classify" title="Link to this heading"></a></h2>
<p>Performs lncRNA classification, classifying RNA sequences as either coding or non-coding.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="o">-</span><span class="n">m</span> <span class="n">lncnrapy</span><span class="o">.</span><span class="n">scripts</span><span class="o">.</span><span class="n">classify</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">model_file</span> <span class="n">MODEL_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">output_file</span> <span class="n">OUTPUT_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">encoding_method</span> <span class="p">{</span><span class="n">cse</span><span class="p">,</span><span class="n">bpe</span><span class="p">,</span><span class="n">kmer</span><span class="p">,</span><span class="n">nuc</span><span class="p">}]</span> <span class="p">[</span><span class="o">--</span><span class="n">bpe_file</span> <span class="n">BPE_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">k</span> <span class="n">K</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">batch_size</span> <span class="n">BATCH_SIZE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">context_length</span> <span class="n">CONTEXT_LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">data_dir</span> <span class="n">DATA_DIR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">results_dir</span> <span class="n">RESULTS_DIR</span><span class="p">]</span> <span class="n">fasta_file</span> <span class="p">[</span><span class="n">fasta_file</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<dl class="simple">
<dt><strong>Positional arguments:</strong></dt><dd><dl class="simple">
<dt><cite>fasta_file</cite></dt><dd><p>Path to FASTA file of RNA sequences or pair of paths to two FASTA files containing protein- and non-coding RNAs, respectively. (str)</p>
</dd>
</dl>
</dd>
<dt><strong>Optional arguments</strong></dt><dd><dl class="simple">
<dt><cite>-h, --help</cite></dt><dd><p>Show help message.</p>
</dd>
<dt><cite>--model_file</cite> MODEL_FILE</dt><dd><p>Trained classifier model, specified by id of a model hosted on the HuggingFace Hub, or a path to a local directory containing model weights. (str=”luukromeijn/lncRNA-BERT-kmer-k3-finetuned”)</p>
</dd>
<dt><cite>--output_file</cite> OUTPUT_FILE</dt><dd><p>Name of .csv/.h5 output file. (str)</p>
</dd>
<dt><cite>--encoding_method</cite> {cse,bpe,kmer,nuc}</dt><dd><p>Sequence encoding method. (str=”kmer”)</p>
</dd>
<dt><cite>--bpe_file</cite> BPE_FILE</dt><dd><p>Filepath to BPE model generated with BPE script. Required when Byte Pair Encoding is used. (str=””)</p>
</dd>
<dt><cite>--k</cite> K</dt><dd><p>Specifies k when K-mer Tokenization is used. (int=3)</p>
</dd>
<dt><cite>--batch_size</cite> BATCH_SIZE</dt><dd><p>Number of samples per prediction step. (int=8)</p>
</dd>
<dt><cite>--context_length</cite> CONTEXT_LENGTH</dt><dd><p>Number of input positions. For cse/k-mer encoding, this translates to a maximum of (768-1)*k input nucleotides. (int=768)</p>
</dd>
<dt><cite>--data_dir</cite> DATA_DIR</dt><dd><p>Parent directory to use for any of the paths specified in these arguments (except for <cite>–model_file</cite>). (str=””)</p>
</dd>
<dt><cite>--results_dir</cite> RESULTS_DIR</dt><dd><p>Parent directory to use for the results folder of this script. (str=””)</p>
</dd>
</dl>
</dd>
</dl>
</section>
<section id="train">
<h2>Train<a class="headerlink" href="#train" title="Link to this heading"></a></h2>
<p>Trains (or fine-tunes) a model (optionally pre-trained) for lncRNA classification.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="o">-</span><span class="n">m</span> <span class="n">lncnrapy</span><span class="o">.</span><span class="n">scripts</span><span class="o">.</span><span class="n">train</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">exp_prefix</span> <span class="n">EXP_PREFIX</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">pretrained_model</span> <span class="n">PRETRAINED_MODEL</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">encoding_method</span> <span class="p">{</span><span class="n">cse</span><span class="p">,</span><span class="n">bpe</span><span class="p">,</span><span class="n">kmer</span><span class="p">,</span><span class="n">nuc</span><span class="p">}]</span> <span class="p">[</span><span class="o">--</span><span class="n">epochs</span> <span class="n">EPOCHS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">n_samples_per_epoch</span> <span class="n">N_SAMPLES_PER_EPOCH</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">batch_size</span> <span class="n">BATCH_SIZE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">learning_rate</span> <span class="n">LEARNING_RATE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">weight_decay</span> <span class="n">WEIGHT_DECAY</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">d_model</span> <span class="n">D_MODEL</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">N</span> <span class="n">N</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">d_ff</span> <span class="n">D_FF</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">h</span> <span class="n">H</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dropout</span> <span class="n">DROPOUT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">hidden_cls_layers</span> <span class="n">HIDDEN_CLS_LAYERS</span> <span class="p">[</span><span class="n">HIDDEN_CLS_LAYERS</span> <span class="o">...</span><span class="p">]]</span> <span class="p">[</span><span class="o">--</span><span class="n">n_kernels</span> <span class="n">N_KERNELS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">kernel_size</span> <span class="n">KERNEL_SIZE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">bpe_file</span> <span class="n">BPE_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">k</span> <span class="n">K</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">context_length</span> <span class="n">CONTEXT_LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">data_dir</span> <span class="n">DATA_DIR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">results_dir</span> <span class="n">RESULTS_DIR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">model_dir</span> <span class="n">MODEL_DIR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">no_weighted_loss</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">no_random_reading_frame</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">freeze_network</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">freeze_kernels</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">input_linear</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">no_input_relu</span><span class="p">]</span> <span class="n">fasta_pcrna_train</span> <span class="n">fasta_ncrna_train</span> <span class="n">fasta_pcrna_valid</span> <span class="n">fasta_ncrna_valid</span>
</pre></div>
</div>
<dl class="simple">
<dt><strong>Positional arguments:</strong></dt><dd><dl class="simple">
<dt><cite>fasta_pcrna_train</cite></dt><dd><p>Path to FASTA file with pcRNA training sequences. (str)</p>
</dd>
<dt><cite>fasta_ncrna_train</cite></dt><dd><p>Path to FASTA file with ncRNA training sequences. (str)</p>
</dd>
<dt><cite>fasta_pcrna_valid</cite></dt><dd><p>Path to FASTA file with pcRNA sequences used for validating the model after every epoch. (str)</p>
</dd>
<dt><cite>fasta_ncrna_valid</cite></dt><dd><p>Path to FASTA file with ncRNA sequences used for validating the model after every epoch. (str)</p>
</dd>
</dl>
</dd>
<dt><strong>Optional arguments</strong></dt><dd><dl class="simple">
<dt><cite>-h, --help</cite></dt><dd><p>Show help message.</p>
</dd>
<dt><cite>--exp_prefix</cite> EXP_PREFIX</dt><dd><p>Added prefix to model/experiment name. (str)</p>
</dd>
<dt><cite>--pretrained_model</cite> PRETRAINED_MODEL</dt><dd><p>If specified, fine-tunes this pre-trained model instead of training one from scratch. Note that this causes model-related hyperparameters, such as d_model and N, to be ignored. Specified by id of a model hosted on the HuggingFace Hub, or a path to a local directory containing model weights. (str)=””</p>
</dd>
<dt><cite>--encoding_method</cite> {cse,bpe,kmer,nuc}</dt><dd><p>Sequence encoding method. (str=”cse”)</p>
</dd>
<dt><cite>--epochs</cite> EPOCHS</dt><dd><p>Number of epochs to train for. (int=100)</p>
</dd>
<dt><cite>--n_samples_per_epoch</cite> N_SAMPLES_PER_EPOCH</dt><dd><p>Number of training samples per epoch. (int=10000)</p>
</dd>
<dt><cite>--batch_size</cite> BATCH_SIZE</dt><dd><p>Number of samples per optimization step. (int=8)</p>
</dd>
<dt><cite>--learning_rate</cite> LEARNING_RATE</dt><dd><p>Learning rate used by Adam optimizer. (float=1e-5)</p>
</dd>
<dt><cite>--weight_decay</cite> WEIGHT_DECAY</dt><dd><p>Weight decay used by Adam optimizer. (float=0.0)</p>
</dd>
<dt><cite>--d_model</cite> D_MODEL</dt><dd><p>BERT embedding dimensionality. (int=768)</p>
</dd>
<dt><cite>--N</cite> N</dt><dd><p>Number of BERT transformer blocks. (int=12)</p>
</dd>
<dt><cite>--d_ff</cite> D_FF</dt><dd><p>Number of nodes in BERT FFN sublayers (int=4*d_model)</p>
</dd>
<dt><cite>--h</cite> H</dt><dd><p>Number of BERT self-attention heads (int=int(d_model/64))</p>
</dd>
<dt><cite>--dropout</cite> DROPOUT</dt><dd><p>Dropout probability in CLS output head. (float=0)</p>
</dd>
<dt><cite>--hidden_cls_layers</cite> HIDDEN_CLS_LAYERS [HIDDEN_CLS_LAYERS …]</dt><dd><p>Space-separated list with number of hidden nodes in ReLU-activated classification head layers. (int=[])</p>
</dd>
<dt><cite>--n_kernels</cite> N_KERNELS</dt><dd><p>Specifies number of kernels when convolutional sequence encoding is used. (int=768)</p>
</dd>
<dt><cite>--kernel_size</cite> KERNEL_SIZE</dt><dd><p>Specifies kernel size when convolutional sequence encoding is used. (int=9)</p>
</dd>
<dt><cite>--bpe_file</cite> BPE_FILE</dt><dd><p>Filepath to BPE model generated with BPE script. Required when Byte Pair Encoding is used. (str=””)</p>
</dd>
<dt><cite>--k</cite> K</dt><dd><p>Specifies k when k-mer encoding is used. (int=6)</p>
</dd>
<dt><cite>--context_length</cite> CONTEXT_LENGTH</dt><dd><p>Number of input positions. For cse/k-mer encoding, this translates to a maximum of (768-1)*k input nucleotides. (int=768)</p>
</dd>
<dt><cite>--data_dir</cite> DATA_DIR</dt><dd><p>Parent directory to use for any of the paths specified in these arguments. (str=””)</p>
</dd>
<dt><cite>--results_dir</cite> RESULTS_DIR</dt><dd><p>Parent directory to use for the results folder of this script. (str=””)</p>
</dd>
<dt><cite>--model_dir</cite> MODEL_DIR</dt><dd><p>Directory where to save the trained model to. Model with highest macro F1-score on the validation dataset is saved. (str=f”{data_dir}/models”)</p>
</dd>
<dt><cite>--no_weighted_loss</cite></dt><dd><p>Applies correction to pcRNA/ncRNA class imbalance. (bool)</p>
</dd>
<dt><cite>--no_random_reading_frame</cite></dt><dd><p>Turns off sampling in random reading frame for convolutional sequence encoding. (bool)</p>
</dd>
<dt><cite>--freeze_network</cite></dt><dd><p>Freezes all weights from the pre-trained model and bases the clasification on the mean embeddings of this model. This only works with the –pretrained_model flag. (bool)</p>
</dd>
<dt><cite>--freeze_kernels</cite></dt><dd><p>Freezes all convolutional sequence encoding weights from the pre-trained model. Only works with the –pretrained_model flag. (bool)</p>
</dd>
<dt><cite>--input_linear</cite></dt><dd><p>Forces linear projection of kernels onto d_model dimensions in convolutional sequence encoding. (bool)</p>
</dd>
<dt><cite>--no_input_relu</cite></dt><dd><p>Turns off ReLU activation of kernels in convolutional sequence encoding. (bool)</p>
</dd>
</dl>
</dd>
</dl>
</section>
<section id="pre-train">
<h2>Pre-Train<a class="headerlink" href="#pre-train" title="Link to this heading"></a></h2>
<p>Pre-training script for a Nucleotide Language Model. Several encoding methods and hyperparameter settings are supported.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="o">-</span><span class="n">m</span> <span class="n">lncnrapy</span><span class="o">.</span><span class="n">scripts</span><span class="o">.</span><span class="n">pretrain</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">exp_prefix</span> <span class="n">EXP_PREFIX</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">encoding_method</span> <span class="p">{</span><span class="n">cse</span><span class="p">,</span><span class="n">bpe</span><span class="p">,</span><span class="n">kmer</span><span class="p">,</span><span class="n">nuc</span><span class="p">}]</span> <span class="p">[</span><span class="o">--</span><span class="n">epochs</span> <span class="n">EPOCHS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">n_samples_per_epoch</span> <span class="n">N_SAMPLES_PER_EPOCH</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">batch_size</span> <span class="n">BATCH_SIZE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">warmup_steps</span> <span class="n">WARMUP_STEPS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">d_model</span> <span class="n">D_MODEL</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">N</span> <span class="n">N</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">d_ff</span> <span class="n">D_FF</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">h</span> <span class="n">H</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">dropout</span> <span class="n">DROPOUT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">n_kernels</span> <span class="n">N_KERNELS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">kernel_size</span> <span class="n">KERNEL_SIZE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">bpe_file</span> <span class="n">BPE_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">k</span> <span class="n">K</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">p_mlm</span> <span class="n">P_MLM</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">p_mask</span> <span class="n">P_MASK</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">p_random</span> <span class="n">P_RANDOM</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">context_length</span> <span class="n">CONTEXT_LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">data_dir</span> <span class="n">DATA_DIR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">results_dir</span> <span class="n">RESULTS_DIR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">model_dir</span> <span class="n">MODEL_DIR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">mask_size</span> <span class="n">MASK_SIZE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">no_random_reading_frame</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">input_linear</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">no_input_relu</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">no_output_linear</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">output_relu</span><span class="p">]</span> <span class="n">fasta_train</span> <span class="n">fasta_valid</span>
</pre></div>
</div>
<dl class="simple">
<dt><strong>Positional arguments:</strong></dt><dd><dl class="simple">
<dt><cite>fasta_train</cite></dt><dd><p>Path to FASTA file with pre-training sequences. (str)</p>
</dd>
<dt><cite>fasta_valid</cite></dt><dd><p>Path to FASTA file with sequences to use for validating model performance after every epoch. (str)</p>
</dd>
</dl>
</dd>
<dt><strong>Optional arguments</strong></dt><dd><dl class="simple">
<dt><cite>-h, --help</cite></dt><dd><p>Show help message.</p>
</dd>
<dt><cite>--exp_prefix</cite> EXP_PREFIX</dt><dd><p>Added prefix to model/experiment name. (str=”MLM”)</p>
</dd>
<dt><cite>--encoding_method</cite> {cse,bpe,kmer,nuc}</dt><dd><p>Sequence encoding method. (str=”cse”)</p>
</dd>
<dt><cite>--epochs</cite> EPOCHS</dt><dd><p>Number of epochs to pre-train for. (int=500)</p>
</dd>
<dt><cite>--n_samples_per_epoch</cite> N_SAMPLES_PER_EPOCH</dt><dd><p>Number of training samples per epoch. (int=10000)</p>
</dd>
<dt><cite>--batch_size</cite> BATCH_SIZE</dt><dd><p>Number of samples per optimization step. (int=8)</p>
</dd>
<dt><cite>--warmup_steps</cite> WARMUP_STEPS</dt><dd><p>Number of optimization steps in which learning rate increases linearly. After this amount of steps, the learning rate decreases proportional to the inverse square root of the step number. (int=8)</p>
</dd>
<dt><cite>--d_model</cite> D_MODEL</dt><dd><p>BERT embedding dimensionality. (int=768)</p>
</dd>
<dt><cite>--N</cite> N</dt><dd><p>Number of BERT transformer blocks. (int=12)</p>
</dd>
<dt><cite>--d_ff</cite> D_FF</dt><dd><p>Number of nodes in BERT FFN sublayers (int=4*d_model)</p>
</dd>
<dt><cite>--h</cite> H</dt><dd><p>Number of BERT self-attention heads (int=int(d_model/64))</p>
</dd>
<dt><cite>--dropout</cite> DROPOUT</dt><dd><p>Dropout probability in MLM output head. (float=0)</p>
</dd>
<dt><cite>--n_kernels</cite> N_KERNELS</dt><dd><p>Specifies number of kernels when convolutional sequence encoding is used. (int=768)</p>
</dd>
<dt><cite>--kernel_size</cite> KERNEL_SIZE</dt><dd><p>Specifies kernel size when convolutional sequence encoding is used. (int=9)</p>
</dd>
<dt><cite>--bpe_file</cite> BPE_FILE</dt><dd><p>Filepath to BPE model generated with BPE script. Required when Byte Pair Encoding is used. (str=””)</p>
</dd>
<dt><cite>--k</cite> K</dt><dd><p>Specifies k when k-mer encoding is used. (int=6)</p>
</dd>
<dt><cite>--p_mlm</cite> P_MLM</dt><dd><p>Selection probability per token/nucleotide in MLM. (float=0.15)</p>
</dd>
<dt><cite>--p_mask</cite> P_MASK</dt><dd><p>Mask probability for selected token/nucleotide. (float=0.8)</p>
</dd>
<dt><cite>--p_random</cite> P_RANDOM</dt><dd><p>Random replacement chance per token/nucleotide. (float=0.1)</p>
</dd>
<dt><cite>--context_length</cite> CONTEXT_LENGTH</dt><dd><p>Number of input positions. For cse/k-mer encoding, this translates to a maximum of (768-1)*k input nucleotides. (int=768)</p>
</dd>
<dt><cite>--data_dir</cite> DATA_DIR</dt><dd><p>Parent directory to use for any of the paths specified in these arguments. (str=””)</p>
</dd>
<dt><cite>--results_dir</cite> RESULTS_DIR</dt><dd><p>Parent directory to use for the results folder of this script. (str=””)</p>
</dd>
<dt><cite>--model_dir</cite> MODEL_DIR</dt><dd><p>Directory where to save pre-trained model to. Model with the highest accuracy on the validation dataset is saved. (str=f”{data_dir}/models”)</p>
</dd>
<dt><cite>--mask_size</cite> MASK_SIZE</dt><dd><p>Number of contiguous nucleotides that make up a mask. (int=1)</p>
</dd>
<dt><cite>--no_random_reading_frame</cite></dt><dd><p>Turns off sampling in random reading frame for convolutional sequence encoding (bool)</p>
</dd>
<dt><cite>--input_linear</cite></dt><dd><p>Forces linear projection of kernels onto d_model dimensions in convolutional sequence encoding. (bool)</p>
</dd>
<dt><cite>--no_input_relu</cite></dt><dd><p>Turns off ReLU activation of kernels in convolutional sequence encoding. (bool)</p>
</dd>
<dt><cite>--no_output_linear</cite></dt><dd><p>Forces linear projection of embeddings onto n_kernels dimensions before masked convolution output layer. (bool)</p>
</dd>
<dt><cite>--output_relu</cite></dt><dd><p>Forces ReLU activation of embeddings before masked convolution output layer. (bool)</p>
</dd>
</dl>
</dd>
</dl>
</section>
<section id="generate-embeddings">
<h2>Generate Embeddings<a class="headerlink" href="#generate-embeddings" title="Link to this heading"></a></h2>
<p>Retrieves sequence embeddings by specified model for input dataset.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="o">-</span><span class="n">m</span> <span class="n">lncnrapy</span><span class="o">.</span><span class="n">scripts</span><span class="o">.</span><span class="n">embeddings</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">model_file</span> <span class="n">MODEL_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">output_file</span> <span class="n">OUTPUT_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">output_plot_file</span> <span class="n">OUTPUT_PLOT_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">encoding_method</span> <span class="p">{</span><span class="n">cse</span><span class="p">,</span><span class="n">bpe</span><span class="p">,</span><span class="n">kmer</span><span class="p">,</span><span class="n">nuc</span><span class="p">}]</span> <span class="p">[</span><span class="o">--</span><span class="n">bpe_file</span> <span class="n">BPE_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">k</span> <span class="n">K</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">pooling</span> <span class="p">{</span><span class="n">CLS</span><span class="p">,</span><span class="n">mean</span><span class="p">,</span><span class="nb">max</span><span class="p">}]</span> <span class="p">[</span><span class="o">--</span><span class="n">dim_red</span> <span class="p">{</span><span class="n">tsne</span><span class="p">,</span><span class="n">pca</span><span class="p">,</span><span class="n">umap</span><span class="p">,</span><span class="kc">None</span><span class="p">}]</span> <span class="p">[</span><span class="o">--</span><span class="n">batch_size</span> <span class="n">BATCH_SIZE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">context_length</span> <span class="n">CONTEXT_LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">data_dir</span> <span class="n">DATA_DIR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">results_dir</span> <span class="n">RESULTS_DIR</span><span class="p">]</span> <span class="n">fasta_file</span> <span class="p">[</span><span class="n">fasta_file</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<dl class="simple">
<dt><strong>Positional arguments:</strong></dt><dd><dl class="simple">
<dt><cite>fasta_file</cite></dt><dd><p>Path to FASTA file of RNA sequences or pair of paths to two FASTA files containing protein- and non-coding RNAs, respectively. (str)</p>
</dd>
</dl>
</dd>
<dt><strong>Optional arguments</strong></dt><dd><dl class="simple">
<dt><cite>-h, --help</cite></dt><dd><p>Show help message.</p>
</dd>
<dt><cite>--model_file</cite> MODEL_FILE</dt><dd><p>(Pre-)trained model, specified by id of a model hosted on the HuggingFace Hub, or a path to a local directory containing model weights. (str=”luukromeijn/lncRNA-BERT-kmer-k3-pretrained”)</p>
</dd>
<dt><cite>--output_file</cite> OUTPUT_FILE</dt><dd><p>Name of .csv/.h5 output file. (str)</p>
</dd>
<dt><cite>--output_plot_file</cite> OUTPUT_PLOT_FILE</dt><dd><p>If specified, plots the first two dimensions of the (reduced) sequence embeddings and saves them to this file. (str)</p>
</dd>
<dt><cite>--encoding_method</cite> {cse,bpe,kmer,nuc}</dt><dd><p>Sequence encoding method. (str=”kmer”)</p>
</dd>
<dt><cite>--bpe_file</cite> BPE_FILE</dt><dd><p>Filepath to BPE model generated with BPE script. Required when Byte Pair Encoding is used. (str=””)</p>
</dd>
<dt><cite>--k</cite> K</dt><dd><p>Specifies k when K-mer Tokenization is used. (int=3)</p>
</dd>
<dt><cite>--pooling</cite> {CLS,mean,max}</dt><dd><p>Type of pooling to apply. If “CLS”, will extract embeddings from CLS token. (str=”mean”)</p>
</dd>
<dt><cite>--dim_red</cite> {tsne,pca,umap,None}</dt><dd><p>Type of dimensionality reduction to apply to retrieved embeddings. If None, will not reduce dimensions. (str=tsne)</p>
</dd>
<dt><cite>--batch_size</cite> BATCH_SIZE</dt><dd><p>Number of samples per prediction step. (int=8)</p>
</dd>
<dt><cite>--context_length</cite> CONTEXT_LENGTH</dt><dd><p>Number of input positions. For cse/k-mer encoding, this translates to a maximum of (768-1)*k input nucleotides. (int=768)</p>
</dd>
<dt><cite>--data_dir</cite> DATA_DIR</dt><dd><p>Parent directory to use for any of the paths specified in these arguments (except for <cite>–model_file</cite>). (str=””)</p>
</dd>
<dt><cite>--results_dir</cite> RESULTS_DIR</dt><dd><p>Parent directory to use for the results folder of this script. (str=””)</p>
</dd>
</dl>
</dd>
</dl>
</section>
<section id="byte-pair-encoding">
<h2>Byte Pair Encoding<a class="headerlink" href="#byte-pair-encoding" title="Link to this heading"></a></h2>
<p>Fits a Byte Pair Encoding (BPE) model to a dataset.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="o">-</span><span class="n">m</span> <span class="n">lncnrapy</span><span class="o">.</span><span class="n">scripts</span><span class="o">.</span><span class="n">bpe</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">bpe_file</span> <span class="n">BPE_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">data_dir</span> <span class="n">DATA_DIR</span><span class="p">]</span> <span class="n">fasta_train</span> <span class="n">vocab_size</span>
</pre></div>
</div>
<dl class="simple">
<dt><strong>Positional arguments:</strong></dt><dd><dl class="simple">
<dt><cite>fasta_train</cite></dt><dd><p>Path to FASTA file of RNA sequences to be used for fitting the BPE model. (str)</p>
</dd>
<dt><cite>vocab_size</cite></dt><dd><p>Pre-defined number of tokens in vocabulary. (str)</p>
</dd>
</dl>
</dd>
<dt><strong>Optional arguments</strong></dt><dd><dl class="simple">
<dt><cite>-h, --help</cite></dt><dd><p>Show help message.</p>
</dd>
<dt><cite>--bpe_file</cite> BPE_FILE</dt><dd><p>Name of BPE output file. (str=f”features/{vocab_size}.bpe”)</p>
</dd>
<dt><cite>--data_dir</cite> DATA_DIR</dt><dd><p>Parent directory to use for any of the paths specified in these arguments. (str=””)</p>
</dd>
</dl>
</dd>
</dl>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="api.html" class="btn btn-neutral float-right" title="API Documentation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Luuk Romeijn.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>